notes/shelley/CommunicationPatterns.md

```diff
@@ -95,6 +150,79 @@ and is missing some that we will need.
     of current subscribers (at the moment these are easily confused in the
     buckets of `Peers`).
 
+## Subscription conversation
+
+There's one conversation per subscription. It's opened by the subscriber and
+initially is subscribed to no topics.
+
+### Subscriber sends
+
+  - AddTopic (Topic msg) : subscribe to a topic.
+  - RemoveTopic (Topic msg) : unsubscribe from a topic.
+  - Request key : get the payload for a key.
+  - Ignore key : signal that we don't want the payload for a key.
+  - Keepalive : send after a certain amount of inactivity. Useful for
+      detecting half-open TCP connections (intermittent lower-layer failures).
+
+### Publisher sends
+
+  - Announce key : indicate that a payload for this key is available.
+  - Payload value : give the payload for a key.
+
+Should the publisher acknowledge add/remove topic?
+If a subscriber adds a topic and then removes it, there may be announces for
+that topic inbound.
```

Comment:
We could send them back with the ack message for remove, or if we don't want them, remove the subscriber only after the queue was cleared up to remove request.  If we have all the requests in a FIFO this property will hold.